### PR TITLE
Add fn groupMembershipInEpoch to validators.sol for slashing

### LINKED ISSUE #2087: Function in Validators to retrieve a validator's group membership at a particular epoch for slashing
### Expected Behavior
```
Contract Validators {
  // Checks the membership history entries of `account` and requires that `epochNumber` 
  // is within the range [entries[index].Epoch, entries[index + 1].Epoch).
  // Returns entries[index].group.
  function groupMembershipInEpoch(address account, uint256 epochNumber, uint256 index) external view returns (address);
}
```

### Current Behavior

No such function

### PR DESCRIPTION
**Pull request by lucasege**
### Description

Added `groupMembershipInEpoch` fn and related testing. Would like review on failure values (currently require(false) on failures). Also a look over logic would be nice.

### Tested

Added unit tests for general usage and edge cases.

### Related issues

- Fixes #2087

### Backwards compatibility

Yes

### CODE REVIEW

File: packages/protocol/contracts/governance/Validators.sol [hunk 1]
```diff
@@ -765,6 +765,7 @@ contract Validators is
       uint256 multiplier = Math.max(1, groups[account].members.numElements);
       uint256[] storage sizeHistory = groups[account].sizeHistory;
       if (sizeHistory.length > 0) {
+        // TODO(lucas): Why isn't this >=
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: What happens when you set it to >=?

File: packages/protocol/contracts/governance/Validators.sol [hunk 2]
```diff
@@ -1131,4 +1132,30 @@ contract Validators is
...
+  * @notice Returns the group that `account` was a member of during `epochNumber`.
+  * @param account The account whose group membership should be returned.
+  * @param epochNumber The epoch number we are querying this account's membership at.
+  * @param index The index into the validator's history struct for their history at `epochNumber`.
+  * @return The group that `account` was a member of during `epochNumber`.
+  */
+  function groupMembershipInEpoch(address account, uint256 epochNumber, uint256 index)
+    external
+    view
+    onlyRegisteredContracts(canForceDeaffiliation)
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Oops :)

File: packages/protocol/contracts/governance/Validators.sol [hunk 3]
```diff
@@ -1131,4 +1132,30 @@ contract Validators is
...
+    MembershipHistory storage history = validators[account].membershipHistory;
+    if (
+      history.entries[index].epochNumber == epochNumber ||
+      (history.entries[index].epochNumber < epochNumber &&
+        (history.entries[index.add(1)].epochNumber > epochNumber ||
+          history.entries[index.add(1)].epochNumber == 0))
+    ) {
+      return history.entries[index].group;
+    }
+    require(false, "provided index does not match provided epochNumber at index in history.");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: The more idiomatic approach would be to require that the condition you have in the `if` holds, and then to return `history.entries[index].group;`

File: packages/protocol/contracts/governance/Validators.sol [hunk 4]
```diff
@@ -1131,4 +1132,30 @@ contract Validators is
...
+    onlyRegisteredContracts(canForceDeaffiliation)
+    returns (address)
+  {
+    require(isValidator(account) && epochNumber <= getEpochNumber());
+    MembershipHistory storage history = validators[account].membershipHistory;
+    if (
+      history.entries[index].epochNumber == epochNumber ||
+      (history.entries[index].epochNumber < epochNumber &&
+        (history.entries[index.add(1)].epochNumber > epochNumber ||
+          history.entries[index.add(1)].epochNumber == 0))
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: I don't think this is strictly correct.

In epoch 3, I join group A (entry index N is A)
In epoch 4, I leave group A (entry index N+1 is 0)
in epoch 5, I join group B (entry index N+2 is B)

groupMembershipInEpoch(me, 5, N) == A 😢

Comment (asaj) [reply packages/protocol/contracts/governance/Validators.sol]: Please add a test case that would have caught this

File: packages/protocol/test/governance/validators.ts [hunk 5]
```diff
@@ -2184,4 +2184,116 @@ contract('Validators', (accounts: string[]) => {
...
+    const validator = accounts[0]
+    const groups = accounts.slice(1, -1)
+
+    beforeEach(async () => {
+      await registry.setAddressFor(CeloContractName.DowntimeSlasher, validator)
+      await registry.setAddressFor(CeloContractName.DoubleSigningSlasher, accounts[3])
+      await registry.setAddressFor(CeloContractName.Governance, accounts[4])
+    })
+
+    describe('when the validator is added to different groups sequantially', async () => {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: typo

File: packages/protocol/contracts/governance/Validators.sol (hunk 1 above)
Comment (lucasege) [reply packages/protocol/contracts/governance/Validators.sol]: Ah cool, interesting "gotcha"

File: packages/protocol/contracts/governance/Validators.sol (hunk 2 above)
Comment (lucasege) [reply packages/protocol/contracts/governance/Validators.sol]: I assume this was an edit?

File: packages/protocol/contracts/governance/Validators.sol (hunk 4 above)
Comment (lucasege) [reply packages/protocol/contracts/governance/Validators.sol]: Addressed, please review again

File: packages/protocol/contracts/governance/Validators.sol (hunk 2 above)
Comment (asaj) [reply packages/protocol/contracts/governance/Validators.sol]: This is a view call, it should be fine for any address to call, so I don't think we need this modifier

File: packages/protocol/contracts/governance/Validators.sol [hunk 6]
```diff
@@ -1131,4 +1131,29 @@ contract Validators is
...
+    onlyRegisteredContracts(canForceDeaffiliation)
+    returns (address)
+  {
+    require(isValidator(account) && epochNumber <= getEpochNumber());
+    MembershipHistory storage history = validators[account].membershipHistory;
+    require(
+      history.entries[index].epochNumber == epochNumber ||
+        (history.entries[index].epochNumber < epochNumber &&
+          (history.entries[index.add(1)].epochNumber > epochNumber ||
+            index == history.numEntries.sub(1))),
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: I don't think this is how you check that this is the last entry in `history` (though I agree that's what you want to check)

File: packages/protocol/contracts/governance/Validators.sol [hunk 7]
```diff
@@ -1131,4 +1131,29 @@ contract Validators is
...
+    external
+    view
+    onlyRegisteredContracts(canForceDeaffiliation)
+    returns (address)
+  {
+    require(isValidator(account) && epochNumber <= getEpochNumber());
+    MembershipHistory storage history = validators[account].membershipHistory;
+    require(
+      history.entries[index].epochNumber == epochNumber ||
+        (history.entries[index].epochNumber < epochNumber &&
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Are there any issues when any of these values are uninitialized? I don't think so, but please make sure you've thought about it carefully

File: packages/protocol/test/governance/validators.ts [hunk 8]
```diff
@@ -2184,4 +2184,129 @@ contract('Validators', (accounts: string[]) => {
       })
     })
   })
+
+  describe('#groupMembershipInEpoch', async () => {
+    const validator = accounts[0]
+    const groups = accounts.slice(1, -1)
+
+    beforeEach(async () => {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Not needed

File: packages/protocol/test/governance/validators.ts [hunk 9]
```diff
@@ -2184,4 +2184,129 @@ contract('Validators', (accounts: string[]) => {
...
+          epochs.push(epochNumber)
+          await mineBlocks(blocksUntilNextEpoch, web3)
+
+          if (i != 4) {
+            await validators.affiliate(groups[i])
+            await validators.addFirstMember(validator, NULL_ADDRESS, NULL_ADDRESS, {
+              from: groups[i],
+            })
+          } else {
+            await validators.forceDeaffiliateIfValidator(validator)
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: You don't need to force deaffiliate, you can just call deaffiliate from the validator address

File: packages/protocol/contracts/governance/Validators.sol (hunk 6 above)
Comment (lucasege) [reply packages/protocol/contracts/governance/Validators.sol]: Right, it should be `index.sub(history.tail) == history.numEntries.sub(1)` since the tail is the smallest index, so `index` - `tail` should give the offset in entries, which if equal to the total entries means we are on the last element. Good catch

File: packages/protocol/contracts/governance/Validators.sol (hunk 7 above)
Comment (lucasege) [reply packages/protocol/contracts/governance/Validators.sol]: I addressed one that I missed which is if the epoch number really was 0, which I think _is possible_. 
I believe all uninitialized cases are handled, assuming history is updated correctly.

File: packages/protocol/test/governance/validators.ts (hunk 8 above)
Comment (lucasege) [reply packages/protocol/test/governance/validators.ts]: I added `await`s before `setAddressFor` to silence linter complaints, so now it is

File: packages/protocol/contracts/governance/Validators.sol (hunk 7 above)
Comment (asaj) [reply packages/protocol/contracts/governance/Validators.sol]: For readability, consider the following:
```
require(index >= history.tail && history.numEntries > 0);
bool isExactMatch = history.entries[index].epochNumber == epochNumber;
bool isLastEntry = index.sub(history.tail) == history.numEntries.sub(1);
bool isWithinRange = history.entries[index].epochNumber < epochNumber && (history.entries[index.add(1)].epochNumber > epochNumber || isLastEntry);
require(isExactMatch || isWithinRange);
```

File: packages/protocol/test/governance/validators.ts (hunk 8 above)
Comment (asaj) [reply packages/protocol/test/governance/validators.ts]: No I mean there's no need to call `setAddressFor` for this test since `groupMembershipInEpoch` can be called by anyone

File: packages/protocol/test/governance/validators.ts [hunk 10]
```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
...
+        await registerValidator(validator)
+        for (const group of groups) {
+          await registerValidatorGroup(group)
+        }
+        for (let i = 0; i < membershipHistoryLength.plus(1).toNumber(); i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+          epochs.push([epochNumber + 1, i !== 4 ? groups[i] : NULL_ADDRESS])
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Why `i !== 4`? Would be nice to include more instances in which the validator was removed from the group, maybe `i % 3 === 0`?

File: packages/protocol/test/governance/validators.ts [hunk 11]
```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
...
+
+          if (i !== 4) {
+            await validators.affiliate(groups[i])
+            await validators.addFirstMember(validator, NULL_ADDRESS, NULL_ADDRESS, {
+              from: groups[i],
+            })
+          } else {
+            await validators.forceDeaffiliateIfValidator(validator)
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: You don't need to force deaffiliate, you can just call deaffiliate from the validator address

File: packages/protocol/test/governance/validators.ts [hunk 12]
```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
...
+          }
+        }
+      })
+
+      it('should always return the correct membership for the given epoch', async () => {
+        for (let i = 0; i < membershipHistoryLength.toNumber(); i++) {
+          assert.equal(
+            await validators.groupMembershipInEpoch(validator, epochs[i + 1][0], 2 + i),
+            i !== 3 ? epochs[i + 1][1] : NULL_ADDRESS
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Please include a comment as to why this is different than above

File: packages/protocol/test/governance/validators.ts [hunk 13]
```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
...
+          } else {
+            await validators.forceDeaffiliateIfValidator(validator)
+          }
+        }
+      })
+
+      it('should always return the correct membership for the given epoch', async () => {
+        for (let i = 0; i < membershipHistoryLength.toNumber(); i++) {
+          assert.equal(
+            await validators.groupMembershipInEpoch(validator, epochs[i + 1][0], 2 + i),
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Can you add more comments explaining these args?

File: packages/protocol/test/governance/validators.ts (hunk 10 above)
Comment (asaj) [reply packages/protocol/test/governance/validators.ts]: Ah, I see you do this below. IMO one test like this is sufficient, what extra coverage does this test give us?

File: packages/protocol/test/governance/validators.ts [hunk 14]
```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
...
+        }
+        for (let i = 0; i < membershipHistoryLength.multipliedBy(3).toNumber(); i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+
+          if (i % 3 === 0) {
+            epochs.push([epochNumber + 1, i !== 9 ? groups[i / 3] : NULL_ADDRESS])
+            if (i !== 9) {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Why do this?

Comment (lucasege) [reply packages/protocol/test/governance/validators.ts]: The reason behind `i % 3 == 0` was to only add `membershipHistoryLength` entries - the reason was to avoid the confusion of only tracking the last `membershipHistoryLength` entries while still adding epochNumber gaps between entries. 
The reason behind `i !== 9` is to add a deaffiliation _between_ other entries. 
I'll definitely change this to be a function of `membershipHistoryLength` rather than magic numbers 3 and 4.
If you want me to do this differently, let me know and I can look into it.

File: packages/protocol/contracts/governance/Validators.sol [hunk 15]
```diff
@@ -1131,4 +1131,30 @@ contract Validators is
...
+   * @return The group that `account` was a member of during `epochNumber`.
+   */
+  function groupMembershipInEpoch(address account, uint256 epochNumber, uint256 index)
+    external
+    view
+    returns (address)
+  {
+    require(isValidator(account) && epochNumber <= getEpochNumber());
+    MembershipHistory storage history = validators[account].membershipHistory;
+    require(index >= history.tail && history.numEntries > 0, "index out of bounds");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol:1219]: Should we check the upper bound as well, i.e. `index < history.tail.add(history.numEntries)`

File: packages/protocol/test/governance/validators.ts [hunk 16]
```diff
@@ -2178,10 +2178,98 @@ contract('Validators', (accounts: string[]) => {
...
+        // epochs stores [epochNumber, group] of the corresponding i + 1 indexed entry on chain
+        // i + 1 because registering groups adds a dummy null address as the first entry
+        epochs = []
+        await registerValidator(validator)
+        for (const group of groups) {
+          await registerValidatorGroup(group)
+        }
+        // Magic gapSize to allow for tests of off by one in both directions
+        const gapSize = 3
+        for (let i = 0; i < membershipHistoryLength.multipliedBy(gapSize).toNumber(); i++) {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Should we go for longer than this, so that we can test that getting the membership for an epoch too far in the past fails?

File: packages/protocol/test/governance/validators.ts [hunk 17]
```diff
@@ -2178,10 +2178,98 @@ contract('Validators', (accounts: string[]) => {
...
+        for (let i = 0; i < membershipHistoryLength.toNumber(); i++) {
+          assert.equal(
+            await validators.groupMembershipInEpoch(validator, epochs[i][0], 1 + i),
+            i !== 3 ? epochs[i][1] : NULL_ADDRESS
+          )
+        }
+      })
+
+      it('should correctly get the group address of epoch numbers between switches', async () => {
+        assert.equal(
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts]: Why not test these conditions in a for loop through over all epochs? As is I find it a bit hard to read

File: packages/protocol/test/governance/validators.ts [hunk 18]
```diff
@@ -2178,10 +2178,98 @@ contract('Validators', (accounts: string[]) => {
...
+          await validators.groupMembershipInEpoch(validator, epochs[4][0] + 1, 5),
+          epochs[4][1]
+        )
+        // `epochs`[2][0] join group A, epoch `epochs`[3][0] deaffiliate, `epochs`[4][0] join group B.
+        // This verifies we don't mistake 0 addresses as the end of history.
+        await assertRevert(validators.groupMembershipInEpoch(validator, epochs[4][0], 3))
+        await assertRevert(validators.groupMembershipInEpoch(validator, epochs[3][0], 3))
+      })
+
+      describe('when called with various malformed inputs', () => {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts:2278]: Can we include tests for bad indexes? i.e. indexes that are either less than or greater than what's covered by `membershipHistory`?

File: packages/protocol/test/governance/validators.ts [hunk 19]
```diff
@@ -2205,13 +2205,116 @@ contract('Validators', (accounts: string[]) => {
...
+
+        it("should revert when epochNumber is greater than the chain's current epochNumber", async () => {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          await assertRevert(
+            validators.groupMembershipInEpoch(validator, epochNumber + 1, contractIndex)
+          )
+        })
+
+        it('should revert when provided index is greater than greatest index on chain', async () => {
```
Comment (asaj) [inline packages/protocol/test/governance/validators.ts:2307]: Should we check that it reverts when the provided index is less than the tail?

**Review by asaj — APPROVED** [review]
Looks great!

File: packages/protocol/test/governance/validators.ts (hunk 19 above)
Comment (lucasege) [reply packages/protocol/test/governance/validators.ts:2307]: Line 2273 checks this by going through all of the indices that have been "pushed out" by adding more than `membershipHistoryLength` members, but for clarity I'll add another test here
